File: Javascript/3.Ottobre/59.SpaceWeekDay5-GoldilocksZone-08Ottobre.js
//todo  Per il quinto giorno della Settimana Spaziale, calcolerai la "zona Riccioli d'Oro" di una stella, ovvero la regione attorno a una stella in cui le condizioni sono "proprio ideali" per la presenza di acqua liquida.

//! Data la massa di una stella, restituisci un array con le distanze di inizio e fine della sua zona Riccioli d'Oro in Unità Astronomiche.
//! Per calcolare la zona Riccioli d'Oro:
//! Trova la luminosità della stella elevando la sua massa alla potenza di 3,5.
//! L'inizio della zona è 0,95 volte la radice quadrata della sua luminosità.
//! La fine della zona è 1,37 volte la radice quadrata della sua luminosità.
//! Restituisci le distanze arrotondate a due cifre decimali.
//! Ad esempio, dato 1 come massa, restituisci [0,95, 1,37].

//? Tempo di esecuzione:
//* 12 Minuti e 02 Secondi

//?Codice scritto senza utilizzo di ChatGpt

function goldilocksZone(mass) {

  let luminosity = mass ** 3.5
  let startStar = (luminosity ** 0.5) * 0.95;
  let endStar = (luminosity ** 0.5) * 1.37;

  let calc = [Number(startStar.toFixed(2)), Number(endStar.toFixed(2))]
  return calc;

}

console.log(goldilocksZone(6))

//?Codice scritto da ChatGpt
//* Non usata


//* ✅ Pro

//* Formula matematica corretta

//* Variabili ben separate e leggibili

//*Arrotondamento pulito a 2 decimali

//* Ritorno chiaro come array [min, max]

//* Codice semplice e diretto

//! ❌ Contro

//! luminosity ** 0.5 ripetuto (può essere ottimizzato)

//! Nessun controllo su input non valido (es. numeri ≤ 0)

//! calc è una variabile evitabile

//? Versione Chat Gpt 

function goldilocksZone(mass) {
  const rootLuminosity = (mass ** 3.5) ** 0.5;

  return [
    Number((rootLuminosity * 0.95).toFixed(2)),
    Number((rootLuminosity * 1.37).toFixed(2))
  ];
}

console.log(goldilocksZone(6));


//todo Considerazioni: 

//! N.B.

File: Python/3.Ottobre/59.Space_week_day5-Goldilocks_zone-08Ottobre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# =========================================================

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

# Per il quinto giorno della Settimana dello Spazio,
# calcolare la zona abitabile di una stella.
#
# Data la massa della stella:
# - luminosità = massa ^ 3.5
# - inizio zona = 0.95 * sqrt(luminosità)
# - fine zona = 1.37 * sqrt(luminosità)
#
# Restituire un array contenente i due valori
# arrotondati a 2 cifre decimali.

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

import math

def goldilocks_zone(mass):
    luminosity_star = mass ** 3.5
    sqrt_luminosity = math.sqrt(luminosity_star)

    start_zone = 0.95 * sqrt_luminosity
    end_zone = 1.37 * sqrt_luminosity

    return [round(start_zone, 2), round(end_zone, 2)]

print(goldilocks_zone(3.7))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

import math

def goldilocks_zone(mass):
    luminosity = mass ** 3.5
    root_luminosity = math.sqrt(luminosity)

    return [
        round(0.95 * root_luminosity, 2),
        round(1.37 * root_luminosity, 2)
    ]

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Soluzione corretta.
# - I passaggi del calcolo sono ben separati.
# - Nomi delle variabili descrittivi.
# - Utilizzo corretto di math.sqrt().
# - Codice facile da leggere e debuggare.

# //! Contro
# - I ";" finali sono inutili in Python.
# - Le variabili start_zone ed end_zone vengono usate una sola volta:
#   possono essere inserite direttamente nel return.
# - luminosity_star è un nome un po' lungo; luminosity è sufficiente.

# =========================================================
# //* Consigli
# =========================================================

# - Continua a spezzare i calcoli in passaggi logici:
#   rende il codice molto leggibile.
# - Evita i ";" finali.
# - Quando una variabile viene utilizzata una sola volta,
#   valuta se inserirla direttamente nel return.
# - Potresti anche usare l'operatore ** 0.5 al posto di
#   math.sqrt(), ma la tua soluzione è più esplicita.

# Valutazione: 10/10
# Soluzione corretta, pulita e ben strutturata.
# Non ci sono errori logici né problemi di efficienza.